### ISSUE TITLE
Build fails with Error code 65 on cordova-ios 4.0.1

### ISSUE THREAD

**Issue by klyonrad** [Telerik-Verified-Plugins/WKWebView#209]
not posting the full error output but this seems to be the important part

    /Users/luka/Dropbox/h_da/2015-02-WS/bachelorTestApp/platforms/ios/BachelorTestApp/Plugins/com.telerik.plugins.wkwebview/MyMainViewController.m:4:9: fatal error: 
    'Cordova/CDVLocalStorage.h' file not found
    #import <Cordova/CDVLocalStorage.h>
        ^
    1 error generated.

Downgrading to cordova-ios 3.9.2 seems to have helped the issue

**Comment by EddyVerbruggen** [comment]
Yeah they made a whole bunch of 'internal' plugins (like `CDVLocalStorage`) 'private', so this plugin can no longer access that class. Should be able to work around it, but I'm wondering if cordova-ios-4 compatibility is needed since Cordova now (with the cordova-ios-4 release) has official support for WKWebView with their own plugin.

Would love to hear everybody's input on this matter.

**Comment by mathiasmuller4sh** [comment]
Hi, They do support WKWebView but there is stil the issue for ajax call from file://
So ionic or angular app are not able to work

**Comment by EddyVerbruggen** [comment]
Oh, that's not very useful then.. unless you use inline angular templates I guess. I'll try and make an effort. The hardest part is making it backward compatible with cordova-ios-3.

**Comment by mathiasmuller4sh** [comment]
Inline template is indeed a nice work around.
Haven't tried it.
I stopped my test when i saw this error:
the Origin null is not allowed by Access-Control-Allow-Origin error on each template..

**Comment by andreialecu** [comment]
Cordova's WKWebView support is iOS9 only. This plugin supports iOS 8 as well.

**Comment by EddyVerbruggen** [comment]
You're best off sticking with cordova-ios-3 for now. cordova-ios has been significantly restructured and a lot of things this plugin needs are now private, so inaccessible.

**Comment by EddyVerbruggen** [comment]
For folks using cordova-ios4 and targeting iOS8 and iOS9 who want to use WKWebView today, there is also [this plugin](https://github.com/apache/cordova-plugins/tree/master/wkwebview-engine-localhost). I didn't test it, but it may be worth a shot.

**Comment by klyonrad** [comment]
>Should be able to work around it, but I'm wondering if cordova-ios-4 compatibility is needed since Cordova now (with the cordova-ios-4 release) has official support for WKWebView with their own plugin.

I personally have also come acrosse the issue [of indexedDB not being available with the official plugin](http://stackoverflow.com/questions/34616896/can-i-use-indexeddb-in-a-cordova-ios-application/34617956).

>there is also this plugin. I didn't test it, but it may be worth a shot.

I would assume that they'll drop support for this one in favor of the new plugin.

So I suggest at least pointing out the compatibility issue in the README of your plugin

**Comment by EddyVerbruggen** [comment]
Good point, added the warning near the top.

**Comment by zb0t** [comment]
Downgrading to 3.9.2 helped. No errors while building. Testing with an Ionic 2 app. Great performance boost. 

Thanks!

**Comment by hussfelt** [comment]
:+1: on 3.9.2! Thanks @zb0t

**Comment by prabhu2003ece** [comment]
Yup downgrading to 3.9.2 works. Looks like 4.0 Version has cordovaLib separates Cordova/CDVLocalStorage.h as private plugin which xcode not including them.

**Comment by EddyVerbruggen** [comment]
Exactly (grrr) and a few other issues as well.

**Comment by manuelpaulo** [comment]
Only one small problem... we need v.4.0 or else Google will kick us of their store:

https://support.google.com/faqs/answer/6325474

**Comment by nitzanwilnai** [comment]
Any update on this?

**Comment by alexkavon** [comment]
> The hardest part is making it backward compatible with cordova-ios-3.

Sounds like a breaking change. Would it be possible to only support cordova-ios-4 with any new versions?

**Comment by alexkavon** [comment]
@zb0t with ios8?

**Comment by zb0t** [comment]
@alexkavon I have tested it with a physical device running 9.2 and simulator. Will try iOS 8 right now and report back

**Comment by zb0t** [comment]
@alexkavon Works like a charm on 8.4 simulator!

**Comment by jpike88** [comment]
Why on Earth are you trying to maintain iOS 8 support? The vast majority of iOS are already on iOS 9 (or can update to it) and it's holding this project back. There are already versions of this plugin that support both, but if you literally cannot support Cordova iOS 4 because of a select few users who can't be bothered to update their devices then I think that's a poor decision.

I'm sticking with [cordova-plugin-wkwebview-engine](https://github.com/apache/cordova-plugin-wkwebview-engine) for now (although the crash recovery on this one looks cool) and have been using it for a few months. It works great and the code is simple. I seriously think you guys should consider ditching iOS 8 support and allowing this plugin to take advantage of Cordova iOS, WKWebView is still cutting edge so why can't this plugin be too?

**Comment by jpike88** [comment]
Better yet, why not approach the Cordova repo guys and ask them to move their repo to Github to facilitate some sort of merge, as this one's crash recovery sounds handy?

**Comment by EddyVerbruggen** [comment]
@zb0t Please also check a real device as AFAIK the simulator is able to load files from the `file://` protocol. That's bitten me when testing WKWebView in the past.

**Comment by zb0t** [comment]
@EddyVerbruggen Okay I will test on a physical device

**Comment by helgetan** [comment]
Any updates on this?

**Comment by zb0t** [comment]
On a physical iOS 9 device it appears to work just fine. Looking for a device with iOS 8!

**Comment by andreialecu** [comment]
There are certain workarounds implemented in this plugin that cordova's doesn't have, and various critical issues in that one as well that are not present in this one due to the use of the local web server. For example this one is critical for my app and I can't swap to that one: https://issues.apache.org/jira/browse/CB-10237?jql=project%20%3D%20CB%20AND%20labels%20%3D%20wkwebview-known-issues

Would be great if there could be a way to update the plugin to work with new cordova versions.

**Comment by andreialecu** [comment]
By the way, another plugin I'm using has figured out how to be compatible with both cordova-ios 3 and 4. It involves reading the `platforms/package.json` file on build using a hook, and depending on the iOS version, changing some macros in its source.

See:
https://github.com/nordnet/cordova-hot-code-push/blob/master/scripts/lib/iosBackwardsCapabilitySupport.js
https://github.com/nordnet/cordova-hot-code-push/blob/master/scripts/afterPrepareHook.js

Unfortunately I don't have time to figure it out myself so I'm leaving it here for someone else.

**Comment by jcesarmobile** [comment]
For those complaining about the cordova plugin (https://github.com/apache/cordova-plugin-wkwebview-engine) and the issues that it has, those are apple bugs.
The way to workaround the bugs is using a local webserver like this plugin does.

So, if you are using the cordova plugin, you just have to install this plugin **too** to add a webserver.
https://github.com/apache/cordova-plugins/tree/master/wkwebview-engine-localhost

After installing **both**, it should work the same way this plugin works, and if you still see bugs, create new issues on http://issues.cordova.io/

**Comment by akofman** [comment]
@EddyVerbruggen Thanks for your awesome job, I'm a fan !

![screen shot 2016-03-31 at 17 19 32](http://38.media.tumblr.com/ebdf7b2618bac59bbb7afeb1fa26308c/tumblr_inline_nzpxi5zD711qzex79_500.gif)

Concerning this plugin I think the official WKWebView is still young and needs time.
So far it didn't consider settings configuration (I did a [PR](https://github.com/apache/cordova-plugin-wkwebview-engine/pull/7) in order to fixe that point).
Also I'm just encountering some issues with the localstorage, and the scroll behaviour looks different. That's what I can see from my side but I guess other developers could find other regressions.

As @jcesarmobile said, using [wkwebview-engine-localhost](https://github.com/apache/cordova-plugins/tree/master/wkwebview-engine-localhost) should help but I don't think it's enough at the moment (from my point of view). And yes it's a very good idea to create issues in order to improve it :+1:

Why I'm rekindling this thread is just because a lot of projects can't be migrated to the official plugin because of regression issues and are just stuck with `cordova-ios@3.9.2`. I think it would better to encourage people to migrate to the latest cordova-ios version and wait for some fixes on the official WKWebView plugin before using it (and it could take time).

So to conclude, that will be awesome if you could fix this issue 😄 
But if you don't have time to do it, I can understand and I'll try to do something.

**Comment by andreialecu** [comment]
I gave the official wkwebview plugin a try and it has so many issues that are resolved in this one that it is unusable.

For example, it is impossible to use `cdvfile://` urls without the local webserver engine plugin, and even then, I could not get `CDVFileTransfer` to work properly, getting various permission errors while trying to save files. I think it tries to be smart and override urls automatically but something doesn't work somewhere along the way and breaks a lot of things.

From my experience attempting this migration, my conclusion is that there are so many problems with the official plugin that it is definitely not usable for any moderately complex app at this point - and I'm not seeing any signs that they're actively maintaining it. There are barely any people using it from what I can see, because there is a severe lack of discussion regarding most of these issues.

That said, however, I think it might be easier to fork the official plugin and add the functionality included in this plugin (specifically the local file system proxy via `http://localhost:12344/Library/`, `/Documents/`, etc), the auto-restart functionality, and the various other tweaks that evolved here.

**Comment by branco91** [comment]
@andreialecu  I recently faced the same issue where the CDVLocalStorage file was failing the build.
I also gave a try to the official wekwebview but ended up with troubles using the FILE_URI from Cordova Camera to Cordova File Transfer.  What I did was to actually use the cdvfille:// with the temp folder
  ` val localPath ="cdvfile://"+imageData.substring(imageData.lastIndexOf("/Users/"),imageData.length)`

(imageData, comes from the Cordova Camera)

This did the trick for me actually.

**Comment by Ramblurr** [comment]
Just starting to use this plugin in an existing project. I was already on ios 4.x. Is there any update regarding 4.x support?

**Comment by barocsi** [comment]
@branco91 can you please elaborate?
Is your solution for Teleriks or the official WKWebview?
Is it to work with images selected from gallery?
Right now every taken picture returns a fileURL not the direct imagedata.
https://github.com/apache/cordova-plugin-camera#module_camera.getPicture suggests using FILE_URI instead of DATA_URL because of crashes.

> NOTE: Photo resolution on newer devices is quite good. Photos selected from the device's gallery are not downscaled to a lower quality, even if a quality parameter is specified. To avoid common memory problems, set Camera.destinationType to FILE_URI rather than DATA_URL.

**Comment by barocsi** [comment]
Is this a sort of 22 catch. Accessing photos requires webserver. Requires this plugin, the other one does not work. However this one never compiles for cordova ios 4. But many other plugins require that. So how others solve this issue while keeping wkwebview support?

**Comment by EddyVerbruggen** [comment]
Can you give examples of plugins requiring cordova-ios-4? That's not a requirement I like to impose on projects.

**Comment by barocsi** [comment]
I have just ran into https://github.com/phonegap/phonegap-plugin-push/releases requires ios 4 above 1.5.3, and is at 1.6.3 already with many essential fixes.

**Comment by apparition47** [comment]
[cordova-plugin-iosrtc](https://github.com/eface2face/cordova-plugin-iosrtc) (WebRTC for iOS) also requires Cordova iOS 4.

**Comment by clauderic** [comment]
For my app, the fix that this plugin provides for `KeyboardDisplayRequiresUserAction` is essential, so moving to the official cordova wkwebview plugin will never be an option unless they implement something of the sort, which I doubt will ever happen.

I also would like to use https://github.com/phonegap/phonegap-plugin-push/, but it requires cordova-ios-4

**Comment by guyromb** [comment]
Is there any milestone planed to fix this issue?

**Comment by dbpieter** [comment]
@clauderic We have exactly the same issue. You can fix KeyboardDisplayRequiresUserAction by implementing the same hacky fix that was implemented in the telerik wkwebview : https://github.com/Telerik-Verified-Plugins/WKWebView/commit/04e8296adeb61f289f9c698045c19b62d080c7e3

**Comment by clauderic** [comment]
Yeah, that's exactly what I ended up doing @dbpieter. Works like a charm.

**Comment by branco91** [comment]
Having the same issue CDVLocalStorage  while compiling, I would like to use the  https://github.com/phonegap/phonegap-plugin-push/, but it requires cordova-ios-4

**Comment by gtzinos** [comment]
Here is the solution guys !!! https://stackoverflow.com/questions/36095819/cordova-ios-error-building-images-xcassets/36968113#36968113

**Comment by Michael-Stanford** [comment]
@EddyVerbruggen "Can you give examples of plugins requiring cordova-ios-4? That's not a requirement I like to impose on projects."

    Installing "phonegap-plugin-push" for ios
    Plugin doesn't support this project's cordova-ios version. cordova-ios: 3.9.2, failed version requirement: >=4.3.0

**Comment by jcesarmobile** [comment]
@Michael-Stanford, he said that 8 months ago, in the meantime Xcode 8 was released requiring  .entitlement files for push notifications and cordova-ios 4.x.x dealt with them, so it's a requirement now

**Comment by Michael-Stanford** [comment]
Thanks for your quick response. The Readme at [https://github.com/Telerik-Verified-Plugins/WKWebView](url) says "At the moment this plugin is NOT compatible with Cordova-iOS-4 (so use 3.x), see #209 for details." Does this mean that this WKWebView is no longer supported?

**Comment by JEGardner** [comment]
The need for me to upgrade to cordova-ios 4 to work with XCode 8 has meant that I've unfortunately had to stop using this plugin on all our apps, which is a real shame as it's great. Had to revert to UIWebView for now as I couldn't get Cordova's version to work with the camera etc.

**Comment by thelamborghinistory** [comment]
This is a kinda sad thread :( bye native transitions. Ionic doesn't cease to disappoint the more you dig into it (not flaming any of the contributor guys of this plugin or anything, just bit disappointed from the overall Ionic experience).

**Comment by phyr0s** [comment]
So.... this plugin will be deprecated forever?

**Comment by PierBover** [comment]
> Ionic doesn't cease to disappoint the more you dig into it (not flaming any of the contributor guys of this plugin or anything, just bit disappointed from the overall Ionic experience).

This is a bug with `WkWebView`. You can blame Apple for nor fixing it:
https://bugs.webkit.org/show_bug.cgi?id=154916

**Comment by schumannd** [comment]
So is this plugin abandoned then? Would be nice to read in the README

**Comment by EddyVerbruggen** [comment]
@schumannd See the comments at the top of the readme, and the issue it refers to. Since then Ionic created its own plugin, we're just keeping this one for older Cordova versions.
